src/util_cs/mod.rs

```diff
@@ -1,3 +1,70 @@
+use crate::LinearCombination;
+use ff::PrimeField;
+
 pub mod bench_cs;
 pub mod metric_cs;
 pub mod test_cs;
+
+pub type Constraint<Scalar> = (
+    LinearCombination<Scalar>,
+    LinearCombination<Scalar>,
+    LinearCombination<Scalar>,
+    String,
+);
+
+pub trait Comparable<Scalar: PrimeField> {
+    fn num_inputs(&self) -> usize;
+    fn num_constraints(&self) -> usize;
+    fn inputs(&self) -> Vec<String>;
+    fn aux(&self) -> Vec<String>;
+    fn constraints(&self) -> Vec<Constraint<Scalar>>;
+
+    fn delta<C: Comparable<Scalar>>(&self, other: &C) -> Delta<Scalar>
+    where
+        Scalar: PrimeField,
+    {
+        let input_count_matches = self.num_inputs() == other.num_inputs();
+        let constraint_count_matches = self.num_constraints() == other.num_constraints();
+
+        let inputs_match = self.inputs() == other.inputs();
+        let constraints_match = self.constraints() == other.constraints();
+
+        let equal =
+            input_count_matches && constraint_count_matches && inputs_match && constraints_match;
+
+        if !input_count_matches {
+            Delta::InputCountMismatch(self.num_inputs(), other.num_inputs())
+        } else if !constraint_count_matches {
+            Delta::ConstraintCountMismatch(self.num_constraints(), other.num_constraints())
+        } else if !constraints_match {
+            let mut mismatch = None;
+
+            let c = self.constraints();
+            let o = other.constraints();
+
+            c.iter().zip(&o).enumerate().for_each(|(i, (a, b))| {
```

Comment:
why not filter?

Reply:
Done, which also makes the implementation do what I intended. (The previous was actually capturing the *last* difference, unintentionally.)